### DISCUSSION TITLE
Having trouble trying to get PB to work with NextJS SSR

### DISCUSSION THREAD

**Discussion by popovidis** [pocketbase/pocketbase#4590] [Q&A] [answered]
Hi there,

I have a server side rendered blog and sitemap which both get data from PB. For building I have a .env.production which gets the deployed databases credentials and should in theory build with the production data. However, I get a 400 Failed to authenticate. I have a function that looks like so but when I put console.log(pb) before returning it nothing returns, I am not sure if this is because console.log does not work when building or because it never runs. 

```js
export const getAdminPocketbase = async () => {
  pb.autoCancellation(false);
  await pb.admins.authWithPassword(constants.PB_EMAIL, constants.PB_PASSWORD);
  return pb;
};
```

Any ideas what could fix this? Thank you!

Additional info: the deployed PB is publicly accessible. I am currently interacting with it with local code.

**Comment by tigawanna** [comment]
where is the new Pocketbase() ?

**Reply by popovidis** [reply]
I have it inside of `/app/lib/pb.js` and from there using `const pb = await getAdminPocketbase()` by importing from there.

**Comment by ganigeorgiev** [comment]
The 400 error indicates that there is maybe something wrong with the credentials, I'm not sure.
Eventually you could check if there is a more detailed error message in the _Admin UI > Logs_.

_Side-note: keep in mind that the `authWithPassword()` call is usually very slow due to the bcrypt hash verification, so if the above is called frequently it is recommended to cache the token and reauthenticate with the password only when it expires. `authRefresh()` is OK to be called multiple times and frequently. The SDK can do that for you if you add the `autoRefreshThreshold` option but it must be out of your async function (technically you don't need an async wrapped function but I guess it depends whether Next.JS and your JS runtime supports top-level await calls):_

```js
await pb.admins.authWithPassword(constants.PB_EMAIL, constants.PB_PASSWORD, {
  // This will trigger auto refresh or auto reauthentication in case
  // the token has expired or is going to expire in the next 30 minutes.
  autoRefreshThreshold: 30 * 60
})
```

**Reply by popovidis** [reply]
![image](https://github.com/pocketbase/pocketbase/assets/120704725/7bb1e727-c81e-43c6-9142-ca5f21a43124)

even with autoRefreshThreshhold I still ended up getting that error.

**Reply by ganigeorgiev** [reply]
The `autoRefreshThreshold` will not fix your error, it will just automatically call `authRefresh()` instead of `authWithPassword()` when there is no need for the latter but again in order to do that you need to move it out of the async function, aka. the above needs to be called only once, usually right after the SDK initialization.

In any case, double check that you are passing the correct `constants.PB_EMAIL`, `constants.PB_PASSWORD` values.

**Reply by popovidis** [reply]
I only call it once outside of async and it should work?

**Reply by ganigeorgiev** [reply]
I mentioned `autoRefreshThreshold` as an optimization, it is not the reason why the authentication fails.

Again, double check that you are sending the correct credentials. If you are not able to use `console.log` or other similar means for debugging, you can try to temporary hardcode the credentials directly in your code just for the test.

**Reply by popovidis** [reply, marked as answer]
I believe I got it working by instead of using .env.local and a .env.production I now have a .env.development and a .env.production, hard coding also worked but ofc not ideal.
